internal/k8s/controllers/layer2_status_controller.go

```diff
@@ -0,0 +1,134 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type StatusFetcher func(namespace, name string) []layer2.IPAdvertisement
+
+type l2StatusEvent struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+}
+
+func (evt *l2StatusEvent) DeepCopyObject() runtime.Object {
+	res := new(l2StatusEvent)
+	res.Name = evt.Name
+	res.Namespace = evt.Namespace
+	return res
+}
+
+func NewL2StatusEvent(namespace, name string) event.GenericEvent {
+	evt := l2StatusEvent{}
+	evt.Name = name
+	evt.Namespace = namespace
+	return event.GenericEvent{Object: &evt}
+}
+
+type Layer2StatusReconciler struct {
+	client.Client
+	Logger   log.Logger
+	NodeName string
+	Chan     <-chan event.GenericEvent
+	// fetch ipAdv object to get interface info
+	StatusFetcher StatusFetcher
+}
+
+func (r *Layer2StatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "Layer2StatusReconciler", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "Layer2StatusReconciler", "end reconcile", req.NamespacedName.String())
+
+	ipAdvS := r.StatusFetcher(req.Namespace, req.Name)
+	if len(ipAdvS) == 0 {
+		err := r.Client.Delete(ctx, &v1beta1.ServiceL2Status{ObjectMeta: metav1.ObjectMeta{Name: req.Name, Namespace: req.Namespace}})
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	state := &v1beta1.ServiceL2Status{ObjectMeta: metav1.ObjectMeta{
+		Name:      req.Name,
+		Namespace: req.Namespace,
+	}}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, state)
+	if err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	desiredState := r.buildDesiredStatus(req.Namespace, req.Name, ipAdvS)
+	if reflect.DeepEqual(state, desiredState) {
+		return ctrl.Result{}, nil
+	}
+
+	// todo: why?
+	// even with this, the cr created first time doesn't contain the latest status
+	state.Status = desiredState.Status
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, state, func() error {
+		state.Labels = desiredState.Labels
+		// not working, the status remains empty unless a manual update
+		state.Status = desiredState.Status
+		return nil
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	state.Status = desiredState.Status
+	err = r.Status().Update(ctx, state)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	level.Debug(r.Logger).Log("controller", "Layer2StatusReconciler", "updated state", dumpResource(state))
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Layer2StatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1beta1.ServiceL2Status{}).
+		WatchesRawSource(&source.Channel{Source: r.Chan}, &handler.EnqueueRequestForObject{}).
+		// only trigger reconcile function when the source object is not a ServiceL2Status
+		// this prevents reconciling cycle because we operate ServiceL2Status itself in the reconcile function
```

Comment:
Yes I got you but in our case I think it's slightly different.
With`ctrl.NewControllerManagedBy(mgr).For().WatchesRawSource()`, both ServiceL2Status and GenericEvent objects can trigger the reconcile and the reconciler itself may create ServiceL2Status object, which leads to a cycle reconciling.
Meanwhile, I have tried deleting the `For()` but the controller stops working.So We have to keep the `For()`  but add an `EventFilter` to filter out ServiceL2Status from triggering the reconciling.

Reply:
I understand, I'm talking about us wanting to reconcile on ServiceL2Status objects to make sure they always match the speaker. For example, if the user (or an unrelated speaker) modifies `status.node`, you'd expect the relevant speaker to override it to the correct value.
Regarding the cycle reconciliation, we can compare the current object with the desired status and updating only if it's needed, that way we won't trigger another reconciliation, e.g in `frrk8s_config_controller.go` we do:
```go
	if reflect.DeepEqual(current.Spec, r.desiredConfiguration.Spec) {
		level.Debug(r.Logger).Log("controller", "FRRK8sReconciler", "event", "not reconciling because of no change")
		return ctrl.Result{}, nil
	}
```

Reply:
> I understand, I'm talking about us wanting to reconcile on ServiceL2Status objects to make sure they always match the speaker. For example, if the user (or an unrelated speaker) modifies `status.node`, you'd expect the relevant speaker to override it to the correct value. 

Indeed.In this case, it seems that I have to remove the EventFilter to allow ServiceL2Status object triggering the reconcile funtion.I'll work on it to see if there is any problem.